ngramoffset.go

```diff
@@ -112,3 +122,211 @@ func (a *arrayNgramOffset) DumpMap() map[ngram]simpleSection {
 func (a *arrayNgramOffset) SizeBytes() int {
 	return 8*len(a.tops) + 4*len(a.bots) + 4*len(a.offsets)
 }
+
+// combinedNgramOffset combines an ascii ngram mapping with a unicode ngram mapping,
+// falling back on unicode for unicode ngrams or ascii ngrams with excessive lengths.
+type combinedNgramOffset struct {
+	asc *asciiNgramOffset
+	uni *arrayNgramOffset
+}
+
+func makeCombinedNgramOffset(ngrams []ngram, offsets []uint32) combinedNgramOffset {
+	// split ngrams & offsets into ascii ngrams and unicode ngrams,
+	// since ascii ngrams can be represented much more compactly (21b instead of 63b)
+	ngramsAscii := make([]uint32, 0, len(ngrams)*11/10)
+	offsetsAscii := make([]uint32, 0, len(ngrams)*11/10)
+
+	ngramsUnicode := make([]ngram, 0, len(ngrams)*2/10)
+	offsetsUnicode := make([]uint32, 0, len(ngrams)*2/10)
+
+	for i, ng := range ngrams {
+		if ng&ngramAsciiMask == ng {
+			ngp := ngramAsciiToPacked(ng)
+			if i == len(ngrams)-1 || ngrams[i+1]&ngramAsciiMask != ngrams[i+1] {
+				// at the end of a section we insert an extra offset with the same ngram,
+				// so the size of the segment can be calculated properly
+				ngramsAscii = append(ngramsAscii, ngp, ngp)
+				offsetsAscii = append(offsetsAscii, offsets[i], offsets[i+1])
+			} else {
+				ngramsAscii = append(ngramsAscii, ngp)
+				offsetsAscii = append(offsetsAscii, offsets[i])
+			}
+			if offsets[i+1]-offsets[i] >= ngramAsciiMaxSectionLength {
```

Comment:
Maybe add a comment why this doesn't out of bounds. If I am not mistaken `len(offsets) == len(ngrams)+1` so we can calculate the length of the last ngram. Is that correct?